### PR TITLE
Cognitive complexity fix

### PR DESCRIPTION
**Pull request by SadCheeto**
## Describe your changes
Recursion and breaks in linear flow have been moved into one singular file. This means these are now method specific meaning we should have a lot less violations. Right now, cognitive complexity is both a metric and a aggregation. Please do write a comment if this should be changed
## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [x] If it is a core feature, I have added thorough tests.
- [x] Have I requested and notified the boys for review of the pull-request?

### CODE REVIEW

File: metrics/cognitive_complexity.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+            result.violations.extend(self._count_cognitive_complexity(file_path))
+        return result
+
+    def _count_cognitive_complexity(self, file_path: Path):
+        violations = []
+        ast: tree_sitter.Tree = self._source_repository.getAst(file_path)
+        lang = self._source_repository.language
+        initial_nodes_query_str = f"""{translate_to[lang]["function"]} @func"""
+        if lang == "c_sharp" or lang == "java":
+            initial_nodes_query_str += f"""{translate_to[lang]["constructor"]} @cons"""
```
Comment (Kobo-coder) [inline metrics/cognitive_complexity.py:34]: As far as i can see this is a concatenation of queries on the same string but is very specific, and would require us to add more checks to this if statement if other languages adhere to same semantics as c# or java.
Couldn't we just swap out two (unfortunately) large queries depending on language, where for c# and java it is the junction of both of these queries?

File: metrics/cognitive_complexity.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        query_breaks = self._source_repository.tree_sitter_language.query(
+            f"""
+            [
+                ({translate_to[self._source_repository.language]["if_statement"]} @if)
+                ({translate_to[self._source_repository.language]["for_statement"]} @for)
+                ({translate_to[self._source_repository.language]["while_statement"]} @for)
+                ({translate_to[self._source_repository.language]["catch_statement"]} @catch)
+                ({translate_to[self._source_repository.language]["break_statement"]} @break)
+                ({translate_to[self._source_repository.language]["continue_statement"]} @continue)
+                {translate_to[self._source_repository.language]["goto"]}
```
Comment (Kobo-coder) [inline metrics/cognitive_complexity.py:68]: Not that I've fully understood the purpose of the `@<keyword>` completely, is it a mistake that there isn't one here?

File: metrics/cognitive_complexity.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+            outer_function_name = self._read_function_name(file_path, node)
+        except ValueError:
+            # pass if there is no identifier, considered faulty tree_sitter parsing
+            return
+        nestedCalls = nestedFunctionCallsQuery.captures(node)
+        for call, _ in nestedCalls:
+            try:
+                name = self._read_function_name(file_path, call)
+            except ValueError:
+                # pass if there is no identifier, considered faulty tree_sitter parsing
```
[the remark this answers is not in the document]
Comment (Kobo-coder) [reply metrics/cognitive_complexity.py:91]: As to not be confused with the same exception but a return statement + same comment on line 84-85

**Review by Kobo-coder — DISMISSED** [review]
Few requested changes otherwise goodie 👍

File: metrics/cognitive_complexity.py (hunk 1 above)
Comment (SadCheeto) [reply metrics/cognitive_complexity.py:34]: After a group discussion, we decided to stick with this and accept that Nested Control Flow does not fit this standard

File: metrics/cognitive_complexity.py (hunk 2 above)
Comment (SadCheeto) [reply metrics/cognitive_complexity.py:68]: I did some smart voodoo witchcraft to get goto to work for an array, but I think we can remove this - will test and upload a commit

**Review by Kobo-coder — APPROVED** [review]

**Pull request merged**